### PR TITLE
Improve the Trash data model

### LINKED ISSUE #43494: Improve the Trash data model
As discussed here:

https://www.notion.so/metabase/Outstanding-Issues-with-the-Trash-0be395258ee74c058dd3df79bf93a5d7?showMoveTo=true&saveParent=true

The short version: with the advent of the Trash, we have separated the concept of "where is this item located in the collection hierarchy" into two parts:

- where do we display the item?, and
- what collection determines whether someone has permission to view/modify an item?

This used to be one thing, but for the Trash, it's two: we need to display the item in the Trash, but permissions are still determined by the location an item used to be in.

My initial implementation used the existing infrastructure (`dashboard.collection_id`, `card.collection_id`, `collection.location`) to determine where we display an item, while adding new columns (`dashboard.trashed_from_collection_id`, `collection.trashed_from_location`) used to check permissions.

But: it's a lower risk change to use the existing infrastructure to check *permissions* (because permissions matter more), while adding new columns to figure out where to display an item. As discussed in the document above, it also neatly avoids some issues that cropped up with our initial testing of the Trash (pagination, the difficulty of migrating existing `archived` items to the Trash, and easy pagination).

I've started implementing this in #42845

### PR DESCRIPTION
**Pull request by johnswanson**
Ok, so I had a realization at the PERFECT time, immediately after the RC cutoff. Great job, brain!

Here's the realization. For the Trash, we need to keep track of two things:

- where the item actually is located in the hierarchy, and

- what collection we should look at to see what permissions apply to the item.

For example, a Card might be in the Trash, but we need to look at Collection 1234 to see that a user has permission to Write that card.

My implementation of this was to add a column,
`trashed_from_collection_id`, so that we could move a Card or a Dashboard to a new `collection_id`, but keep track of the permissions we actually needed to check.

So:

- `collection_id` was where the item was located in the collection hierarchy, and

- `trashed_from_collection_id` was where we needed to look to check permissions.

Today I had the realization that it's much, much more important to get PERMISSIONS right than to get collection hierarchy right. Like if we mess up and show something as in the Trash when it's not in the Trash, or show something in the wrong Collection - that's not great, sure. But if we mess up and show a Card when we shouldn't, or show a Dashboard when we shouldn't, that's Super Duper Bad.

So the problem with my initial implementation was that we needed to change everywhere that checked permissions, to make sure they checked BOTH `trashed_from_collection_id` and `collection_id` as appropriate.

So... there's a much better solution. Instead of adding a column to represent the *permissions* that we should apply to the dashboard or card, add a column to represent the *location in the hierarchy* that should apply to the dashboard or the card.

We can simplify further: the *only time* we want to display something in a different place in the hierarchy than usual is when it was put directly into the trash. If you trash a dashboard as a part of a collection, then we should display it in that collection just like normal.

So, we can do the following:

- add a `trashed_directly` column to Cards and Dashboards, representing whether they should be displayed in the Trash instead of their actual parent collection

- use the `collection_id` column of Cards and Dashboards without modification to represent permissions.

There's one main downside of this approach. If you trash a dashboard, and then delete the collection that the dashboard was originally in, what do we do with `dashboard.collection_id`?

- we have to change it, because it's a foreign key

- we can't set it to null, because that represents the root collection

In this initial implementation, I've just cascaded the delete: if you delete a dashboard and then delete a collection, the dashboard will be deleted. This is not ideal. I'm not totally sure what we should do in this situation.

Fixes #43494

### CODE REVIEW

File: src/metabase/models/collection.clj
Comment (johnswanson) [inline src/metabase/models/collection.clj:391]: There was an existing bug here. When looking for collection children, we didn't care at all about `:archived`. So with collections like `/A/B/C` where B is archived, we'd show A as having no effective children.

I'm a little concerned about performance here and still thinking about whether we can help that.

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (rafpaf) [inline frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:67]: I know this line was not added in this PR, but this could use some `ttag.c` love to clarify that it's a verb not a noun

**Review by rafpaf — APPROVED** [review]
FE side looks good

File: src/metabase/models/collection.clj
Comment (noahmoss) [inline src/metabase/models/collection.clj:446]: This is never true, right? Because we already check if `collection-ids` is `:all` on L439 and then shadow the `collection-ids` var with the results of a t2 call

**Review by noahmoss — COMMENTED** [review]
Went through roughly half, will continue tomorrow!

File: resources/migrations/trash/h2.sql
Comment (johnswanson) [reply resources/migrations/trash/h2.sql:9]: Yes, that's right. One thing to note is that we're doing a kind of "best effort" thing here - we can't really know for sure whether something was trashed directly or not. For example, maybe someone trashed a dashboard, then later on trashed the collection it used to be in. We'll mark it as `trashed_directly: false` when it wasn't.

I think that's ok/inevitable, but just wanted to call it out. The knowledge of what was `trashed_directly` has been lost because we didn't record it at the moment it happened. We just do the best we can to do something reasonable.

File: src/metabase/api/card.clj
Comment (johnswanson) [reply src/metabase/api/card.clj:198]: Good call, I'll rename this.

File: src/metabase/api/collection.clj
Comment (johnswanson) [reply src/metabase/api/collection.clj:108]: Yeah :/

Initially I wanted to rename everything to do with `archive` to `trash`. However, I didn't want to hassle with additional migrations and the code churn required to actually change the column names (or the API) over. So that inconsistency is just there - although honestly, maybe it would be better to just use the word "Trash" on the frontend, and have the backend consistently refer to "The archive" or "archived" items. 🤔 

Then, in this case, it's worse, because I let the inconsistent column name bleed into more code. So really, I suppose this should be `include-trashed` rather than `include-archived`... even though it's checking the `:archived` column? 😭 

To answer your question about why we need both `include-archived` and `include-trash?` - `include-trash?` is whether we should include the Trash Collection itself, and `include-archived` is whether we should include things that are "in" the trash.

File: src/metabase/api/collection.clj
Comment (johnswanson) [reply src/metabase/api/collection.clj:529]: Ooh, I wonder. It does seem like if JDBC interprets `bit(1)` as a boolean we should be able to do away with this.

I can check after the other PR is merged.

File: src/metabase/models/collection.clj
Comment (johnswanson) [reply src/metabase/models/collection.clj:303]: Good call, thanks - will take that out.

File: src/metabase/models/collection.clj
Comment (johnswanson) [reply src/metabase/models/collection.clj:446]: Yep, good call - thanks.

File: src/metabase/api/collection.clj
Comment (johnswanson) [reply src/metabase/api/collection.clj:529]: Though I did add `trashed_directly` as a `${boolean.type}` in the migrations. So something odd is going on. 🤔

File: src/metabase/api/collection.clj
Comment (noahmoss) [reply src/metabase/api/collection.clj:108]: I'm personally in favor of consistent naming even if it requires migrations/a little more code churn. Otherwise we built up a bank of implicit historical knowledge you need to figure out to approach the codebase. 

But if it's too much to do as part of this PR, maybe open a tech debt issue for it? At least it'll be documented and could be a good onboarding task in the future.

For the naming of these flags, maybe `include-archived-items?` and `include-trash-collection?`?

File: src/metabase/models/collection.clj
Comment (noahmoss) [inline src/metabase/models/collection.clj:812]: Should this be in the transaction?

File: src/metabase/models/collection.clj
Comment (noahmoss) [inline src/metabase/models/collection.clj:829]: There are a couple places now in this namespace where we enumerate the "models which can live in a collection" which makes me worried that it'll be easy to miss updating these if a new model is added. 

Maybe this could be pulled out as a standalone var? I kind of like the concept of `collectable-models`. Then you'd only need to hard-code Card and Dashboard in these functions as the ones that can be trashed directly (or that could be a separate var)

File: src/metabase/models/collection.clj
Comment (noahmoss) [inline src/metabase/models/collection.clj:1636]: Why not just have the `:can_delete` hydration detect whether the item is a collection? Otherwise we're relying on the caller to use the correct hydration method

File: src/metabase/search/impl.clj
Comment (noahmoss) [inline src/metabase/search/impl.clj:556]: this was my own code, but I just realized that we don't need the `when` since if `effective_parent` is nil it'll be ignored by the merge 😀 
```suggestion
                                  ;; for  non-root collections, override :collection with the values for its effective parent
                                  effective_parent
```

**Review by noahmoss — APPROVED** [review]
Finished going through, looks good to me overall! Most comments are minor so I'm stamping to unblock, but I'm happy to look again at any parts that change

File: src/metabase/models/collection.clj
Comment (johnswanson) [reply src/metabase/models/collection.clj:1636]: Oh, that's much better - thank you!

File: src/metabase/api/card.clj
Comment (johnswanson) [reply src/metabase/api/card.clj]: Renamed this to `api.common/present-in-trash-if-archived-directly` (the column was also renamed from `trashed_directly` to `archived_directly`)

File: src/metabase/api/collection.clj
Comment (johnswanson) [reply src/metabase/api/collection.clj]: Renamed to `:include-archived-items` and `:include-trash-collection?`, and replaced `trashed` with `archived` everywhere, including in migrations (e.g. `trashed_directly` => `archived_directly` and `trash_operation_id` => `archive_operation_id`)

File: src/metabase/api/common.clj
Comment (johnswanson) [reply src/metabase/api/common.clj]: This seems like a good idea - I've:
- [added a *request-id* dynamic var to config](https://github.com/metabase/metabase/pull/42845/files/9ca1e84e50d057636c6fd3f4711daa638b3526de..4cf0c4cd4f25bf6c3aec6acc5297375cbad1dd21#diff-a02bbaeb1685f3f566f2913949ecf046bf8507f757dcc6bd0359ee89bd12d0c4R163-R165)
- [bound that in a new middleware](https://github.com/metabase/metabase/pull/42845/files/9ca1e84e50d057636c6fd3f4711daa638b3526de..4cf0c4cd4f25bf6c3aec6acc5297375cbad1dd21#diff-90cbda8884d3b0752f52adb405bfb073b552341b2860cf210f2753c5a888186fR1-R9)
- [used that](https://github.com/metabase/metabase/pull/42845/files/9ca1e84e50d057636c6fd3f4711daa638b3526de..4cf0c4cd4f25bf6c3aec6acc5297375cbad1dd21#diff-ddea8b16dfd87ccdb499efcc15deca578afe6982b7bba016c75510b6bc484529R315-R328) as the cache args for `collection-id->collection`

File: src/metabase/models/collection.clj
Comment (johnswanson) [reply src/metabase/models/collection.clj]: Yeah, that makes sense - done!

**Pull request merged**
